profiler/profiler.go

```diff
@@ -200,43 +199,14 @@ func newProfiler(opts ...Option) (*profiler, error) {
 		}
 	}
 	p.uploadFunc = p.upload
-	p.telemetry = &telemetry.Client{
-		APIKey:    cfg.apiKey,
-		Namespace: telemetry.NamespaceProfilers,
-		Service:   cfg.service,
-		Env:       cfg.env,
-		Client:    p.cfg.httpClient, // use the profiler's http.Client, gives us UDS if that's used
-	}
-
-	// For the initial release, prefer off-by-default rather than
-	// on-by-default
-	if !internal.BoolEnv("DD_INSTRUMENTATION_TELEMETRY_ENABLED", false) {
-		p.telemetry.Disabled = true
-	}
-
-	// For the URL, uploading through agent goes through
-	//	${AGENT_URL}/telemetry/proxy/api/v2/apmtelemetry
-	// for agentless (which we technically don't support):
-	//	https://instrumentation-telemetry-intake.datadoghq.com/api/v2/apmtelemetry
-	// with an API key
-	//
-	// TODO: move this logic into the telemetry package, make the URL field
-	// unnecessary?
-	if cfg.agentless {
-		p.telemetry.URL = "https://instrumentation-telemetry-intake.datadoghq.com/api/v2/apmtelemetry"
-	} else {
-		// TODO: check agent /info endpoint to see if the agent is
-		// sufficiently recent to support this endpiont? overkill?
-		u, err := url.Parse(cfg.agentURL)
-		if err == nil {
-			u.Path = "/telemetry/proxy/api/v2/apmtelemetry"
-			p.telemetry.URL = u.String()
-		} else {
-			log.Warn("Agent URL %s is invalid, not starting telemetry", cfg.agentURL)
-			p.telemetry.Disabled = true
-		}
-	}
-
+	p.telemetry = telemetry.NewClient(
```

Comment:
Does this mean we'd have two clients running when a user has profiling and tracing enabled?

Reply:
Yes, but there is a `Namespace` field for the client that can be set to `profiler`, `tracer`, or `appsec` to differentiate them

Reply:
> Yes, but there is a Namespace field for the client that can be set to profiler, tracer, or appsec to differentiate them

I double-checked, and it looks like this only gets used for the metrics messages.

When I originally implemented the client, I thought that it would be intended to have tracing, profiling, and appsec all be able to send telemetry. Tracing and profiling can be used independently, so I think it would make sense to be able to enable it in both places. But now that I'm looking through the telemetry docs again, it's less clear to me how it's supposed to be done...

Reply:
AFAIK, it is unexpected from the telemetry API (3 times a service running, 3 times the amount of heart-beats, and so on), on top of being inefficient for us 🤔

Unless what I just said is wrong, I was rather imagining this package to become a singleton package shared by all the packages of dd-trace-go. WDYT?

Reply:
> Unless what I just said is wrong, I was rather imagining this package to become a singleton package shared by all the packages of dd-trace-go. WDYT?

Yeah, looking through the docs again it seems like telemetry might be intended to use that way? It's confusing to me though, because for example tracing and profiling have different configuration. But there's one `app-started` message, and it's only supposed to be sent once per "tracer" lifecycle... the tracer and profiler would need to coordinate to send their configuration values together, I guess, and we'd need a way to distinguish the config options later if we want to look at the data?

I'm wondering now if it would be best to _not_ use this for the profiler, and only send telemetry from the tracer... In fact this might clarify why there is a `profiler` part of the `product` field I mentioned in the other comment.

Reply:
That's why the singleton solves the issue. Something like a simple `telemetry.Instance()` for example. But the multi-tenant configuration becomes harder to design for sure because we'd have to do everything at `init()` time.

Reply:
Right. It seems that I misunderstood how this would work back when I wrote the client. How about for now we delete the telemetry stuff from the profiler. We can revisit how to send data from the profiler later.

Reply:
Got it - will remove the telemetry client from the profiler for now